### PR TITLE
get secret from master based on user/password auth

### PR DESCRIPTION
**Pull request by Constantin07**
This PR is to address the issue in #75 .

It's  doesn't break existing functionality (use of only JENKINS_SECRET).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -26,7 +26,9 @@ Optional environment variables:
 
 * `JENKINS_URL`: url for the Jenkins server, can be used as a replacement to `-url` option, or to set alternate jenkins URL
 * `JENKINS_TUNNEL`: (`HOST:PORT`) connect to this agent host and port instead of Jenkins server, assuming this one do route TCP traffic to Jenkins master. Useful when when Jenkins runs behind a load balancer, reverse proxy, etc.
-* `JENKINS_SECRET`: agent secret, if not set as an argument
+* `JENKINS_SECRET`: agent secret, if not set as an argument (mutually exclusive to `JENKINS_USERNAME` and `JENKINS_PASSWORD`)
+* `JENKINS_USERNAME`: username to authenticate to master (used to dynamicly get the secret from master)
+* `JENKINS_PASSWORD`: password to authenticate to master (used to dynamicly get the secret from master)
```
Comment (oleg-nenashev) [inline README.md]:
```suggestion
* `JENKINS_PASSWORD`: password or API token to authenticate to master (used to dynamically get the secret from master)
```

File: jenkins-slave [hunk 2]
```diff
@@ -29,6 +29,22 @@
 # * JENKINS_SECRET : agent secret, if not set as an argument
 # * JENKINS_AGENT_NAME : agent name, if not set as an argument
 # * JENKINS_AGENT_WORKDIR : agent work directory, if not set by optional parameter -workDir
+# * JENKINS_USERNAME : username to use when authenticating to master -credentials
+# * JENKINS_PASSWORD : password to use when authenticating to master -credentials
```
Comment (oleg-nenashev) [inline jenkins-slave]:
```suggestion
# * JENKINS_PASSWORD : password or API token to use when authenticating to master -credentials
```

File: jenkins-slave [hunk 3]
```diff
@@ -29,6 +29,22 @@
...
+
+
+# Get the secret from Jenkins master
+get_secret() {
+	local jenkins_url="$1"
+	local jenkins_agent_name="$2"
+	local username="$3"
+	local password="$4"
+	local crumb=".crumb:dac7ce5614f8cb32a6ce75153aaf2398"
```
Comment (oleg-nenashev) [inline jenkins-slave]: This CRUMB should be retrieved dynamically for instances

File: README.md (hunk 1 above)
Comment (Constantin07) [reply README.md]: Updated as suggested.

File: jenkins-slave (hunk 2 above)
Comment (Constantin07) [reply jenkins-slave]: Updated as suggested.

File: jenkins-slave (hunk 3 above)
Comment (Constantin07) [reply jenkins-slave]: Updated to get the crumb dynamically. Thanks for the tip @oleg-nenashev.

Note: Interestingly, just tested several times and it works just fine even with hardcoded crumb despite CSRF is properly configured on master.

**Comment by Constantin07** [comment]
@oleg-nenashev done.

**Comment by syndbg** [comment]
@Constantin07 Great work!

@oleg-nenashev It would be great if this gets merged and published in DockerHub.

**Comment by jglick** [comment]
This seems like a bad idea. You are replacing a limited token good only for connecting this agent with a general personal access token which, if compromised, could be abused in various ways. What exactly is the use case here?

**Comment by jglick** [comment]
The use case I can think of is that you are configuring the controller via JCasC including a static agent, and wish to be able to reconstruct the setup from scratch (with a freshly generated `$JENKINS_HOME/secrets/` directory). But then I think you would still need to use custom scripting, at least pending https://github.com/jenkinsci/configuration-as-code-plugin/issues/1830, in which case you may as well script the retrieval of the agent HMAC as well.

The better solution would be to deprecate the agent HMAC generally. Instead each newly created agent would get a randomly generated connection token persisted as a (`Secret`) field in the agent definition; from JCasC you could set a specific token value drawn from the usual sources (e.g., Vault). @daniel-beck @Wadeck do you know if this idea is tracked anywhere?

**Comment by Wadeck** [comment]
=> https://issues.jenkins.io/browse/JENKINS-70168 just created
It was discussed inside SECURITY tracker but it should be public.

**Pull request closed without merging**